### PR TITLE
Cache constants for fetched EVP_cipher

### PR DESCRIPTION
**Pull request by slontis**
##### Checklist

- [ ] documentation is added or updated
- [ ] tests are added or updated

### CODE REVIEW

File: crypto/evp/evp_lib.c [hunk 1]
```diff
@@ -272,16 +272,38 @@ int EVP_CIPHER_type(const EVP_CIPHER *ctx)
...
+
+    params[0] = OSSL_PARAM_construct_size_t(OSSL_CIPHER_PARAM_BLOCK_SIZE, &blksz);
+    params[1] = OSSL_PARAM_construct_size_t(OSSL_CIPHER_PARAM_IVLEN, &ivlen);
+    params[2] = OSSL_PARAM_construct_size_t(OSSL_CIPHER_PARAM_KEYLEN, &keylen);
+    params[3] = OSSL_PARAM_construct_uint(OSSL_CIPHER_PARAM_MODE, &mode);
+    params[4] = OSSL_PARAM_construct_ulong(OSSL_CIPHER_PARAM_FLAGS, &flags);
+    params[5] = OSSL_PARAM_construct_end();
     ok = evp_do_ciph_getparams(cipher, params);
+    if (ok) {
+        /* Provided implementations may have a custom cipher_cipher */
```
Comment (richsalz) [inline crypto/evp/evp_lib.c]: provider?

**Comment by richsalz** [comment]
much gain for few lines of code; nice!

**Comment by beldmit** [comment]
Is it safe enough to cache flags? I'm not familiar with the provider code so are the cached values cached per implementation?

**Comment by slontis** [comment]
The values that are cached here are equivalent to values that were in the old const EVP_CIPHER structs.. i.e- they are constant default values that should not change. The flags should be treated in the same way as the key_length (i.e this is the constant default key length).

**Comment by beldmit** [comment]
I'm not sure about flags because some of them can be implementation-defined (e.g. `EVP_CIPH_FLAG_FIPS`).

**Comment by slontis** [comment]
The values have been stored into the EVP_cipher during the fetch operation as they need to exist before either the new_ctx method or the ctx_int have been called.

e.g- it needs to handle the sequence
ciph = EVP_CIPHER_fetch(.....)
len = EVP_CIPHER_key_length(ciph)

**Comment by slontis** [comment]
> I'm not sure about flags because some of them can be implementation-defined (e.g. `EVP_CIPH_FLAG_FIPS`).

Well that particular one is a dead flag left over from the old fips module.

File: crypto/evp/evp_lib.c (hunk 1 above)
Comment (levitte) [reply crypto/evp/evp_lib.c]: Either works.  The implementations are provided by a provider.

File: crypto/evp/evp_lib.c [hunk 2]
```diff
@@ -272,16 +272,38 @@ int EVP_CIPHER_type(const EVP_CIPHER *ctx)
...
+        if (cipher->prov != NULL && cipher->ccipher != NULL)
+            flags |= EVP_CIPH_FLAG_CUSTOM_CIPHER;
```
Comment (levitte) [inline crypto/evp/evp_lib.c]: Why is this necessary.  `EVP_CIPH_FLAG_CUSTOM_CIPHER` is only looked at in legacy code, and the place where we do use `ccipher`, we check for its existance, not this flag:

https://github.com/openssl/openssl/blob/bdbf2df2e685ae653f3c683ce2f734eb0c0888e0/crypto/evp/evp_lib.c#L312-L316

Comment (slontis) [reply crypto/evp/evp_lib.c]: Happy to take it out,
It was added in
https://github.com/openssl/openssl/commit/f7397f0d58ce7ddf4c5366cd1846f16b341fbe43
The documentation for EVP_cipher that was changed in this commit seems to not be correct now.. It said it looks for the flag (it doesnt seem to though??)

Comment (levitte) [reply crypto/evp/evp_lib.c]: "historical semantics"...  oh, right, that.

No, *we* don't check because we don't need to.  However, callers may want to know what semantics to expect, and the only recourse *they* have is that flag.

[runs in circles, screaming in horror]

Comment (levitte) [reply crypto/evp/evp_lib.c]: So yeah, let's leave this alone.

**Comment by paulidale** [comment]
@t8m, I don't think it changes the semantics quite like that: a cipher that doesn't know/support one of the cached parameters won't return it from the get query and it will default to zero.  True, the _unsupported_ return code won't happen anymore.

It would be an easy enough change to keep the _unsupported_ return code, if that is preferable.

I like the idea of all ciphers supporting all of these constants.  I do wonder what the IV length for an ECB mode cipher should be.  Unsupported seems more reasonable than zero.

**Comment by levitte** [comment]
Why do you want to distinguish an unsupported IV length from zero? Is there a practical reason?

**Comment by paulidale** [comment]
It's the first example of a change that sprung to mind -- agreed, it isn't a good one :sleeping:

This does represent a (small) change of behaviour which is the issue.

**Comment by levitte** [comment]
Actually, this gets us back to pre-3.0 behavior, no?

**Comment by paulidale** [comment]
@levitte, did you intend to approve this?  It is still missing an OMC approval....

**Comment by levitte** [comment]
> @levitte, did you intend to approve this? It is still missing an OMC approval....

I was too quick on the labeling...  but since you approved, all's good.

**Comment by p-steuer** [comment]
merged.

master: https://github.com/openssl/openssl/commit/3c957bcd54d097167e53660fa100aa1ba85d63b5
